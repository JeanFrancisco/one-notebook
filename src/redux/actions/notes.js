import Swal from 'sweetalert2';
import { addDoc, collection, deleteDoc, doc, getDocs, orderBy, updateDoc } from '@firebase/firestore';
import {
    NOTES_NEW_NOTE,
    NOTES_SET_ACTIVE,
    NOTES_DISPLAY_IN_SIDEBAR,
    NOTES_UPDATED_NOTE,
    NOTES_DROP_NOTE,
    NOTES_CLEANING_WHEN_LOGOUT,
    NOTES_CHANGE_ORDER,
} from '../constants/typesNotes';
import { uploadFile } from '../../helpers/files';
import { db as firestoreDB } from '../../config/firebase';

export const actionAddNewNote = () => {
    return async (dispatch, getState) => {
        const { uid: user_uid } = getState().auth;

        const newNote = {
            title: '',
            body: '',
            urlImage: '',
            createdAt: new Date().getTime(),
        };

        const doc = await addDoc( collection(firestoreDB, `${user_uid}/journal/notes`), newNote);

        const createdNote = {
            id: doc.id,
            ...newNote,
        };

        dispatch( actionNewNote( createdNote ) );
        dispatch( actionSetActiveNote( createdNote ) );
    }
};

export const actionNewNote = ( createdNote ) => ({
    type: NOTES_NEW_NOTE,
    payload: createdNote,
});

export const actionSetActiveNote = ( activeNote ) => ({
    type: NOTES_SET_ACTIVE,
    payload: activeNote,
});

export const actionSavingChanges = () => {
    return async (dispatch, getState) => {
        const user_uid = getState().auth.uid;
        const activeNote = getState().notes.active;

        const { id: note_id, ...noteForFirestore } = activeNote;
        await updateDoc( doc(firestoreDB, `${user_uid}/journal/notes/${ note_id }`), noteForFirestore )
            .then(() => {
                dispatch( refreshNote(activeNote) );

                Swal.fire({
                    title: 'Guardada exitosamente',
                    html: '<p><small>Su nota se guardó correctamente</small></p>',
                    icon: 'success',
                    timer: 1300,
                    showConfirmButton: true,
                    confirmButtonText: 'Aceptar',
                    confirmButtonColor: 'limegreen',
                });
            })
            .catch(reason => {
                console.log( reason.code );
            });
    }
}

export const refreshNote = ( updatedNote ) => ({
    type: NOTES_UPDATED_NOTE,
    payload: { ...updatedNote },
});

export const getFirebaseNotes = () => {
    return async ( dispatch, getState ) => {
        const { uid } = getState().auth;

        const snapNotes = await getDocs( collection(firestoreDB, `${uid}/journal/notes`) );
        const notes = [];

        snapNotes.forEach( snapHijo => {
            notes.push({
                id: snapHijo.id,
                ...snapHijo.data(),
            })
        });

        dispatch( displayNotes(notes) );
    }
}

export const displayNotes = ( notes ) => ({
    type: NOTES_DISPLAY_IN_SIDEBAR,
    payload: notes,
});

export const actionSetAscendingOrder = () => ({
    type: NOTES_CHANGE_ORDER,
    payload: {
        order: 'asc',
    }
});

export const actionSetDescendingOrder = () => ({
    type: NOTES_CHANGE_ORDER,
    payload: {
        order: 'desc',
    }
});

export const actionReorderByRule = ( orderBy ) => ({
    type: NOTES_CHANGE_ORDER,
    payload: {
        orderby: orderBy,
    }
});

export const actionStartUploadFile = (file) => {
    return async (dispatch, getState) => {
 
        Swal.fire({
            title: 'Subiendo archivos...',
            text: 'Por favor, espere un momento...',
            allowOutsideClick: false,
            showConfirmButton: false,
            willOpen: () => {
                Swal.showLoading();
            },
        });

        const activeNote = getState().notes.active;
        const { secure_url } = await uploadFile(file);

        dispatch( actionSetActiveNote({ ...activeNote, urlImage: secure_url }) );
        await dispatch( actionSavingChanges() );

        Swal.close();
    }
}

export const actionStartDeletingNote = (note_id) => {
    return async (dispatch, getState) => {
        const user_uid = getState().auth.uid;

        try {
            const resp = await deleteDoc( doc( firestoreDB, `${user_uid}/journal/notes/${note_id}`) );

            dispatch( deleteNote( note_id ) );

            Swal.fire({
                icon: 'success',
                timer: 1250,
                title: 'Eliminado',
                text: 'Su nota fue eliminada',
                confirmButtonText: 'Aceptar',
                confirmButtonColor: 'green',
                showCloseButton: true,
            });

        } catch (e) {
            Swal.fire({
                icon: 'error',
                html: 'Error en la conexión con Firestore DB',
                title: 'Error. <small>No pudimos terminar la eliminación</small>',
                footer: `<small>For more details: ${e.code}</small>`,
                confirmButtonText: 'Entendido',
                confirmButtonColor: 'oragered', 
                showCloseButton: true,
            });

            console.error(e);
        }
    }
}

export const deleteNote = (noteID) => ({
    type: NOTES_DROP_NOTE,
    payload: noteID
});

export const cleanupNotesLogout = () => ({
    type: NOTES_CLEANING_WHEN_LOGOUT
});