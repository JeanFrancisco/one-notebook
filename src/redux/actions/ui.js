import { UI_FINISH_LOADING, UI_START_LOADING } from "../constants/typesUi";

export const actionStartLoading = () => ({
    type: UI_START_LOADING,
});

export const actionFinishLoading = () => ({
    type: UI_FINISH_LOADING,
});