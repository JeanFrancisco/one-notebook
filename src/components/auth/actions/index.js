import {
    createUserWithEmailAndPassword,
    signInWithEmailAndPassword,
    GoogleAuthProvider,
    signInWithPopup,
    updateProfile,
    getAuth,
    signOut,
} from "firebase/auth";
import Swal from 'sweetalert2';
import { app } from "../../../config/firebase";
import { cleanupNotesLogout } from "../../../redux/actions/notes";
import { actionFinishLoading, actionStartLoading } from "../../../redux/actions/ui";
import {
    AUTH_LOGIN,
    AUTH_LOGOUT,
    AUTH_SET_ERROR,
    AUTH_REMOVE_ERROR
} from "../constants";

export const authLogin = (uid, displayName) => ({
    type: AUTH_LOGIN,
    payload: {
        uid,
        displayName,
    }
});

export const authSignInWithGoogle = () => {
    return async (dispatch) => {

        dispatch( actionStartLoading() );

        const auth = getAuth( app );

        await signInWithPopup( auth, new GoogleAuthProvider() )
        .then( userCredentials => {
            // const { user: { displayName, uid } } = userCredentials;
            
            // dispatch( authLogin(uid, displayName) );
        })
        .catch( reason => {
            Swal.fire({
                icon: 'warning',
                html: '<small>No se pudo iniciar la sesión. <br/> <br/> ¿Alcanzó a ver si ocurrió algo: con el Pop-up de autenticación de Google?</small>',
                title: 'Oups! Lo sentimos.',
                iconHtml: '<i class="fa fa-warning"></i>',
                footer: `<small>Reason Code: ${reason.code}</small>`,
                showCloseButton: true,
                confirmButtonText: 'Aceptar',
            });
        })
        .finally(() => {
            dispatch( actionFinishLoading() );
        });
    };
}

export const authEmailPasswordSignIn = ({ email, password }) => {
    return async dispatch => {
        dispatch( actionStartLoading() );

        const auth = getAuth(app);

        await signInWithEmailAndPassword(auth, email, password)
            .then( ({ user }) => {
                // const { uid, displayName } = user;

                // dispatch( authLogin(uid, displayName) );
            }).catch( reason => {
                Swal.fire({
                    title: 'Oups! <small>Algo ocurrió mientras iniciabamos tu sesión.</small>',
                    html: `<p><b><small>¿Qué acciones puedes hacer?</small></b></p>
                        <ul style="text-align: left">
                            <li><small>Puedes revisar por favor tu email y contraseña.</small></li>
                            <li><small>También que te puedes conectar a Internet. En especial a los servicios de Google.</small></li>
                        </ul>`,
                    confirmButtonText: 'Entendido',
                    icon: 'warning',
                    iconColor: 'orange',
                    showCloseButton: true,
                    confirmButtonColor: 'orangered',
                    footer: `<small>Reason Code: ${reason.code}</small>`,
                })
            }).finally(() => {
                dispatch( actionFinishLoading() );
            });
    }
}

export const authEmailPasswordSignUp = ({ email, password, username }) => {
    return async (dispatch) => {
        dispatch( actionStartLoading() );

        const auth = getAuth(app);

        await createUserWithEmailAndPassword(auth, email, password)
            .then( async userCredentials => {

                const { user } = userCredentials;
                await updateProfile(user, { displayName: username });

                // dispatch( authLogin(user.uid, user.displayName) );

            }).catch( reason => {
                Swal.fire({
                    title: 'Oups! <small>Ahora mismo no podemos registrarte.</small>',
                    html: '<p>Por favor... <br/> <small>Regresa más tarde y vuelve a intentarlo ;) .</small><p>',
                    icon: 'info',
                    iconColor: 'orange',
                    confirmButtonText: 'Entiendo',
                    confirmButtonColor: 'limegreen',
                    showCloseButton: true,
                    footer: `<small>Reason Code: ${reason.code}</small>`,
                });
            }).finally(() => {
               dispatch( actionFinishLoading() );
            });
    }
}

export const authFirebaseLogout = () => {
    return async dispatch => {
        const auth = getAuth(app);

        await signOut(auth);

        dispatch( authLogout() );

        dispatch( cleanupNotesLogout() );
    }
}

export const authLogout = () => ({ type: AUTH_LOGOUT });

export const setAuthError = ( msg ) => ({
    type: AUTH_SET_ERROR,
    payload: msg
});

export const removeAuthError = () => ({ type: AUTH_REMOVE_ERROR });