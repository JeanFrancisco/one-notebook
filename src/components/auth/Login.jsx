import React, { useEffect } from 'react';
import { Link } from 'react-router-dom';
import { useDispatch, useSelector } from 'react-redux';
import validator from 'validator';
import Errors from './Errors';
import { useForm } from '../../hooks/useForm';
import { authSignInWithGoogle, authEmailPasswordSignIn, removeAuthError, setAuthError } from './actions';

const Login = () => {
    const dispatch = useDispatch();
    const { errors } = useSelector( store => store.auth );
    const { loading } = useSelector( store => store.ui );

    // useEffect(() => {
    //     dispatch( removeAuthError() );
    // }, [ dispatch ]);

    const isFormValid = () => {
        if( validator.isEmpty( formValues.email, ) || validator.isEmpty( formValues.password ) ) {
            dispatch( setAuthError('Por favor, asegúrese de ingresar todos los campos.'))
            return false;
        }

        if( ! validator.isEmail( formValues.email ) ) {
            dispatch( setAuthError('Por favor asegúrese que su email es correcto, el valor ingresado no es válido.') );
            return false;
        }

        if( ! validator.isLength( formValues.password, { min: 6 } )) {
            dispatch( setAuthError('La contraseña es incorrecta, como mínimo debe tener una longitud de 6 caracteres.') );
            return false;
        }

        if( errors !== undefined )
            dispatch( removeAuthError() );

        return true;
    }

    const onSubmit = () => {
        if( isFormValid() ) {
            const action = authEmailPasswordSignIn( formValues );
            dispatch( action );
        }
    }

    const [ formValues, handleInputChanges, handleFormSubmit ] = useForm({
        email: '',
        password: ''
    }, onSubmit)

    const handleSignInWithGoogle = e => {
        const asyncAction = authSignInWithGoogle();
        dispatch( asyncAction );
    }

    return <>
        <h4 className="auth__title">Iniciar Sesión</h4>

        <Errors errors={ errors } />

        <form onSubmit={ handleFormSubmit }>
            <input
                type="email"
                name="email"
                className="auth__input"
                autoComplete="off"
                value={ formValues.email }
                onChange={ handleInputChanges }
                placeholder="Ingresa aquí tu e-mail."/>

            <input
                type="password"
                name="password"
                className="auth__input"
                value={ formValues.password }
                onChange={ handleInputChanges }
                placeholder="Escribe aquí tu contraseña."
                />

            <button
                className="btn btn-primary btn-block auth__button"
                disabled={ loading }
                type="submit">
                    Ingresar
            </button>

            <div className="auth__social-networks">
                <p>Iniciar sesión con redes sociales:</p>

                <div
                    onClick={ handleSignInWithGoogle }
                    disabled={ loading }
                    className="auth__google-btn">
                    <div className="auth__google-icon-wrapper">
                        <img
                            className="auth__google-icon"
                            src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg"
                            alt="[google logo]" />
                    </div>
                    <div className="auth__google-btn-text">
                        <b>Sign in with google</b>
                    </div>
                </div>
            </div>
        </form>

        <Link
            className="link"
            to="/auth/register">
            Crear una cuenta nueva.
        </Link>
    </>;
}

export default Login;