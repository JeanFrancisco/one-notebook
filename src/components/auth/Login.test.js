import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import Login from './Login';
import { authSignInWithGoogle, authEmailPasswordSignIn, removeAuthError, setAuthError } from './actions/index';

const mockStore = configureStore([ thunk ]);

jest.mock('./actions/index', () => ({
    setAuthError: jest.fn().mockReturnValue({}),
    removeAuthError: jest.fn().mockReturnValue({}),
    authSignInWithGoogle: jest.fn(),
    authEmailPasswordSignIn: jest.fn(),
}));

let store, wrapper;

describe('Pruebas al componente <Login/>.jsx', () => {

    beforeEach(() => {
        store = mockStore({
            auth: {
                errors: undefined,
            },
            ui: {
                loading: false,
            }
        });
        
        store.dispatch = jest.fn();
        
        wrapper = mount(<Provider store={ store }>
            <MemoryRouter>
                <Login />
            </MemoryRouter>
        </Provider>);        
    });

    test('Debe hacer match con el snapshot por defecto', () => {
        const store = mockStore({
            auth: {
                errors: undefined,
            },
            ui: {
                loading: false,
            }
        });

        const wrapper = mount(<Provider store={ store }>
            <MemoryRouter>
                <Login />
            </MemoryRouter>
        </Provider>);

        expect( wrapper ).toMatchSnapshot();
    });

    test('Debe mandar a llamar la función|acción de signInWithGoogle(), al dar click en el botón submit', () => {
        wrapper.find('.auth__google-btn').simulate('click');

        expect( authSignInWithGoogle ).toHaveBeenCalled();
    });

    test('Debe llamar a la funcion setAuthError() si algún input no tiene información correcta', () => {
        wrapper.find('form').prop('onSubmit')({ preventDefault() {} });

        expect( setAuthError ).toHaveBeenCalledWith('Por favor, asegúrese de ingresar todos los campos.');

        wrapper.find('input[name="email"]').simulate('change', {
            target: {
                name: 'email',
                value: 'jane_doe_example.org',
            }
        });

        wrapper.find('input[name="password"]').simulate('change', { target: { name: 'password', value: '012' } });

        wrapper.find('form').prop('onSubmit')({ preventDefault() {} });

        expect( setAuthError ).toHaveBeenCalledWith('Por favor asegúrese que su email es correcto, el valor ingresado no es válido.');

        wrapper.find('input[name="email"]').simulate('change', {
            target: {
                name: 'email',
                value: 'jane_doe@example.org',
            }
        });

        wrapper.find('form').prop('onSubmit')({ preventDefault() {} });

        expect( setAuthError ).toHaveBeenCalledWith('La contraseña es incorrecta, como mínimo debe tener una longitud de 6 caracteres.');
    });

    test('Debe llamar a la acción authEmailPasswordSignIn() con los parámetros correspondientes', () => {
        wrapper.find('input[name="email"]').simulate('change', {
            target: {
                name: 'email',
                value: 'jane_doe@example.org',
            }
        });

        wrapper.find('input[name="password"]').simulate('change', { target: { name: 'password', value: '092187438' } });

        wrapper.find('form').simulate('submit', { preventDefault(){} });

        expect( authEmailPasswordSignIn ).toHaveBeenCalledWith({
            password: '092187438',
            email: 'jane_doe@example.org'
        });
    });
});