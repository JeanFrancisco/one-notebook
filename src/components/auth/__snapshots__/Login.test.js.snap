// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas al componente <Login/>.jsx Debe hacer match con el snapshot por defecto 1`] = `
Array [
  <h4
    className="auth__title"
  >
    Iniciar Sesión
  </h4>,
  "",
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="auth__input"
      name="email"
      onChange={[Function]}
      placeholder="Ingresa aquí tu e-mail."
      type="email"
      value=""
    />
    <input
      className="auth__input"
      name="password"
      onChange={[Function]}
      placeholder="Escribe aquí tu contraseña."
      type="password"
      value=""
    />
    <button
      className="btn btn-primary btn-block auth__button"
      disabled={false}
      type="submit"
    >
      Ingresar
    </button>
    <div
      className="auth__social-networks"
    >
      <p>
        Iniciar sesión con redes sociales:
      </p>
      <div
        className="auth__google-btn"
        disabled={false}
        onClick={[Function]}
      >
        <div
          className="auth__google-icon-wrapper"
        >
          <img
            alt="[google logo]"
            className="auth__google-icon"
            src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg"
          />
        </div>
        <div
          className="auth__google-btn-text"
        >
          <b>
            Sign in with google
          </b>
        </div>
      </div>
    </div>
  </form>,
  <a
    className="link"
    href="/auth/register"
    onClick={[Function]}
  >
    Crear una cuenta nueva.
  </a>,
]
`;
